pkg/demoinfocs/s2_commands.go

```diff
@@ -0,0 +1,345 @@
+	msgs2.ETEProtobufIds_TE_BubblesId:        func() proto.Message { return &msgs2.CMsgTEBubbles{} },
+	msgs2.ETEProtobufIds_TE_BubbleTrailId:    func() proto.Message { return &msgs2.CMsgTEBubbleTrail{} },
+	msgs2.ETEProtobufIds_TE_DecalId:          func() proto.Message { return &msgs2.CMsgTEDecal{} },
+	msgs2.ETEProtobufIds_TE_WorldDecalId:     func() proto.Message { return &msgs2.CMsgTEWorldDecal{} },
+	msgs2.ETEProtobufIds_TE_EnergySplashId:   func() proto.Message { return &msgs2.CMsgTEEnergySplash{} },
+	msgs2.ETEProtobufIds_TE_FizzId:           func() proto.Message { return &msgs2.CMsgTEFizz{} },
+	msgs2.ETEProtobufIds_TE_ShatterSurfaceId: func() proto.Message { return &msgs2.CMsgTEShatterSurface{} },
+	msgs2.ETEProtobufIds_TE_GlowSpriteId:     func() proto.Message { return &msgs2.CMsgTEGlowSprite{} },
+	msgs2.ETEProtobufIds_TE_ImpactId:         func() proto.Message { return &msgs2.CMsgTEImpact{} },
+	msgs2.ETEProtobufIds_TE_MuzzleFlashId:    func() proto.Message { return &msgs2.CMsgTEMuzzleFlash{} },
+	msgs2.ETEProtobufIds_TE_BloodStreamId:    func() proto.Message { return &msgs2.CMsgTEBloodStream{} },
+	msgs2.ETEProtobufIds_TE_ExplosionId:      func() proto.Message { return &msgs2.CMsgTEExplosion{} },
+	msgs2.ETEProtobufIds_TE_DustId:           func() proto.Message { return &msgs2.CMsgTEDust{} },
+	msgs2.ETEProtobufIds_TE_LargeFunnelId:    func() proto.Message { return &msgs2.CMsgTELargeFunnel{} },
+	msgs2.ETEProtobufIds_TE_SparksId:         func() proto.Message { return &msgs2.CMsgTESparks{} },
+	msgs2.ETEProtobufIds_TE_PhysicsPropId:    func() proto.Message { return &msgs2.CMsgTEPhysicsProp{} },
+	msgs2.ETEProtobufIds_TE_PlayerDecalId:    func() proto.Message { return &msgs2.CMsgTEPlayerDecal{} },
+	msgs2.ETEProtobufIds_TE_ProjectedDecalId: func() proto.Message { return &msgs2.CMsgTEProjectedDecal{} },
+	msgs2.ETEProtobufIds_TE_SmokeId:          func() proto.Message { return &msgs2.CMsgTESmoke{} },
+}
+
+type pendingMessage struct {
+	t   int32
+	buf []byte
+}
+
+// Calculates the priority of the message. Lower is more important.
+func (m *pendingMessage) priority() int {
+	switch m.t {
+	case
+		// These messages provide context needed for the rest of the tick
+		// and should have the highest priority.
+		int32(msgs2.NET_Messages_net_Tick),
+		int32(msgs2.SVC_Messages_svc_CreateStringTable),
+		int32(msgs2.SVC_Messages_svc_UpdateStringTable),
+		int32(msgs2.NET_Messages_net_SpawnGroup_Load):
+		return -10
+
+	case
+		// These messages benefit from having context but may also need to
+		// provide context in terms of delta updates.
+		int32(msgs2.SVC_Messages_svc_PacketEntities):
+		return 5
+
+	}
+
+	return 0
+}
+
+func (p *parser) handleDemoPacket(pack *msgs2.CDemoPacket) {
+	r := bitread.NewSmallBitReader(bytes.NewReader(pack.Data))
+
+	ms := make([]pendingMessage, 0)
+
+	for r.ActualPosition() < len(pack.Data) {
+		t := int32(r.ReadUBitInt())
```

Comment:
If I understood correctly, this one starts reading the buffer in 6 bits increments. Is it supposed to advance the offset in non-byte increments?

Reply:
The reading is not "byte aligned", you just read bits from the stream. Also ReadUBitInt reads a varying number of bits, not always 6 bits.